framed.go

```diff
@@ -158,28 +251,56 @@ func (framed *Writer) Write(frame []byte) (n int, err error) {
 	return
 }
 
-func (framed *Writer) WritePieces(pieces ...[]byte) (n int, err error) {
-	framed.mutex.Lock()
-	defer framed.mutex.Unlock()
+// WriteAtomic writes a the frame and its length header in a single write. This requires
+// that the frame was read into a buffer obtained from a pool created with
+// NewHeaderPreservingBufferPool().
+func (fr *Writer) WriteAtomic(frame bpool.ByteSlice) (n int, err error) {
+	if !fr.threadSafetyDisabled {
+		fr.mutex.Lock()
+		defer fr.mutex.Unlock()
+	}
 
-	for _, piece := range pieces {
-		n = n + len(piece)
+	n = len(frame.Bytes())
+	_frame := frame.Full()
+
+	switch fr.bigFramesEnabled {
+	case true:
+		endianness.PutUint32(_frame, uint32(n))
+	default:
+		endianness.PutUint16(_frame, uint16(n))
 	}
 
-	if n > MaxFrameLength {
-		return 0, fmt.Errorf(tooLongError, n, MaxFrameLength)
+	// Write frame and data atomically
+	var written int
+	if written, err = fr.Stream.Write(_frame); err != nil {
+		return
+	}
+	expectedToWrite := n + fr.headerLength
+	if written != expectedToWrite {
+		err = fmt.Errorf("%d bytes written, expected to write %d", written, expectedToWrite)
+	}
```

Comment:
Nit: io.Writers are supposed to return a non-nil error if they don't write the full input.

So it should be sufficient to do:
```
n, err := fr.Stream.Write(_frame)
return max(n - headerLen, 0), err
```

This doesn't actually matter, just simplifies things.

Reply:
Cool, will simplify.